src/AWS.Deploy.Common/IO/DirectoryManager.cs

```diff
@@ -1,23 +1,94 @@
 // Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 // SPDX-License-Identifier: Apache-2.0
 
+using System;
 using System.IO;
 
 namespace AWS.Deploy.Common.IO
 {
     public interface IDirectoryManager
     {
         DirectoryInfo CreateDirectory(string path);
+        DirectoryInfo GetDirectoryInfo(string path);
         bool Exists(string path);
-        string[] GetFiles(string projectPath, string? searchPattern = null);
+        string[] GetFiles(string path, string? searchPattern = null, SearchOption searchOption = SearchOption.TopDirectoryOnly);
+        string[] GetDirectories(string path, string? searchPattern = null, SearchOption searchOption = SearchOption.TopDirectoryOnly);
+        bool IsEmpty(string path);
+        bool ExistsInsideDirectory(string parentDirectoryPath, string childPath);
+        void Delete(string path, bool recursive = false);
     }
 
     public class DirectoryManager : IDirectoryManager
     {
-        public DirectoryInfo CreateDirectory(string path) => Directory.CreateDirectory(path);
+        public DirectoryInfo CreateDirectory(string path)
+        {
+            try
+            {
+                return Directory.CreateDirectory(path);
+            }
+            catch (Exception ex)
+            {
+                throw new FailedToCreateDirectoryException($"Failed to create a directory at path - {path}", ex);
+            } 
+        }
```

Comment:
File IO wrappers are for mocking only. We must not add any logic over the static methods.

We should capture the right exception in the caller.

Reply:
There is different File I/O wrapper called `TestDirectoryManager` and `TestFileManager` used for mocking. Adding the exception in the wrapper class saves the effort into wrapping **every** call around a try-catch block.

Reply:
> File IO wrappers are for mocking only.

I wasn't clear enough, for mocking I meant, converting static methods to instance methods by wrapping them in `*Manager` class. 

Depending on the use case, we should use better messaging than just "Failed to do something". This information is already supplied by the exceptions thrown in FileIO methods, [one example](https://docs.microsoft.com/en-us/dotnet/api/system.io.directory.createdirectory?view=net-5.0).

I expect, if generate command fails, we must write something

```
Failed to generate deployment project

UnAuthorizedAccessException: Unable to access file. The file is read-only. // This is one example
```

Reply:
The invocation of the generate command is wrapped around a try-catch block inside `CommandFactory.cs`. If the diagnostics mode is turned on then the entire stack trace is printed. But without it only the top-level exception is printed.

The custom exceptions are also catching any exception thrown by System.IO and storing them as inner exceptions.